src/join-9.sol

```diff
@@ -35,65 +33,84 @@ interface GemLike {
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
     modifier auth {
-        require(wards[msg.sender] == 1, "GemJoin/not-authorized");
+        require(wards[msg.sender] == 1, "GemJoin9/not-authorized");
         _;
     }
 
-    VatLike public vat;   // CDP Engine
-    bytes32 public ilk;   // Collateral Type
-    GemLike public gem;
-    uint256 public dec;
-    uint256 public live;  // Active Flag
+    VatLike public immutable vat;   // CDP Engine
+    bytes32 public immutable ilk;   // Collateral Type
+    GemLike public immutable gem;
+    uint256 public immutable dec;
+    uint256 public live;            // Active Flag
     uint256 public total;
 
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Join(address indexed usr, uint256 wad);
+    event Exit(address indexed usr, uint256 wad);
+    event Cage();
+
     constructor(address vat_, bytes32 ilk_, address gem_) public {
-        wards[msg.sender] = 1;
-        live = 1;
         vat = VatLike(vat_);
         ilk = ilk_;
         gem = GemLike(gem_);
-        dec = gem.decimals();
+        dec = GemLike(gem_).decimals();
+        live = 1;
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
     }
-    function cage() external note auth {
+    function cage() external auth {
         live = 0;
+        emit Cage();
     }
+
     function add(uint256 x, uint256 y) internal pure returns (uint256 z) {
-        require((z = x + y) >= x);
+        require((z = x + y) >= x, "GemJoin9/overflow");
     }
     function sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
-        require((z = x - y) <= x);
+        require((z = x - y) <= x, "GemJoin9/underflow");
     }
+
     // Allow dss-proxy-actions to send the gems with only 1 transfer
     // This should be called via token.transfer() followed by gemJoin.join() atomically or
     // someone else can steal your tokens
-    function join(address usr) public note returns (uint256) {
-        require(live == 1, "GemJoin/not-live");
+    function join(address usr) public returns (uint256) {
+        require(live == 1, "GemJoin9/not-live");
 
         uint256 _total = total;     // Cache to save an SLOAD
         uint256 wad = sub(gem.balanceOf(address(this)), _total);
-        require(int256(wad) >= 0, "GemJoin/overflow");
+        require(int256(wad) >= 0, "GemJoin9/overflow");
 
         vat.slip(ilk, usr, int256(wad));
         total = add(_total, wad);
 
         return wad;
     }
     function join(address usr, uint256 wad) external {
-        require(gem.transferFrom(msg.sender, address(this), wad), "GemJoin/failed-transfer");
+        require(gem.transferFrom(msg.sender, address(this), wad), "GemJoin9/failed-transfer");
         join(usr);
+        emit Join(usr, wad);
```

Comment:
Hmm I don't like it, there is no event when accessing to the other `join` function.
I'd suggest to extract the functionality to an internal and have each of the `join` as external printing the respective event.

Reply:
done in fb4bcdd.